### PR TITLE
WIP: Added Read/Write Validation for Tables

### PR DESCRIPTION
**Pull request by SanchayGupta1197**
#### Pull Request Description
Basic PR
- Added initial validation for read and write access (IT test to be added)
- Changed UserName, Password, APP key, Authority, AppId to ```Type.Password``` to prevent them from leaking into the logs. in KustoSinkConfig
- Added Kusto.sink.auto_table_create config in KustoSinkConfig.java

### CODE REVIEW

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 1]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+        } catch (URISyntaxException e) {
+            throw new ConnectException("Unable to connect to Server", e);
+        } catch (DataClientException e) {
+            e.printStackTrace();
+        } catch (DataServiceException e) {
+            if (e.getCause().getMessage().contains("Database")) {
+                throw new ConfigException(
+                        String.format("Invalid or Denied Access to Database: %s", database), e);
+            } else if (e.getCause().getMessage().contains("Table")) {
+                if (config.getKustoAutoTableCreate()) {
```
Comment (fahadhasher) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:211]: this should be done when you try getting the schema of the table(line:178). Try creating a table depending on the `KustoAutoTableCreate` flag and fail with a message `Unable to create %s table in Kusto DB.` if the operation fails or else proceed.

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 2]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+            String extentId = rs.getValues().get(0).get(0);
+            //Deleting the temporary record.
+            engineClient.execute(database, String.format(".drop extent %s", extentId));
+            log.info("User has read and Write access to the table {}", table);
+        } catch (URISyntaxException e) {
+            throw new ConnectException("Unable to connect to Server", e);
+        } catch (DataClientException e) {
+            e.printStackTrace();
+        } catch (DataServiceException e) {
+            if (e.getCause().getMessage().contains("Database")) {
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:207]: I had but them as a config exception as there might e a case where the database passed in the config file doesn't exist(due to a typo).

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkConfig.java [hunk 3]
```diff
@@ -26,6 +26,7 @@
     static final String KUSTO_SINK_FLUSH_SIZE = "kusto.sink.flush_size";
     static final String KUSTO_SINK_FLUSH_INTERVAL_MS = "kusto.sink.flush_interval_ms";
     static final String KUSTO_SINK_WRITE_TO_FILES = "kusto.sink.write_to_files";
+    static final String KUSTO_AUTO_TABLE_CREATE = "kusto.auto_table_create";
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkConfig.java:29]: done

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkConfig.java [hunk 4]
```diff
@@ -88,5 +90,7 @@ public String getKustoSinkTempDir() {
     public long getKustoFlushIntervalMS() {
         return this.getLong(KUSTO_SINK_FLUSH_INTERVAL_MS);
     }
+
+    public boolean getKustoAutoTableCreate() {return this.getBoolean(KUSTO_AUTO_TABLE_CREATE); }
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkConfig.java:94]: done,

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkConfig.java [hunk 5]
```diff
@@ -46,7 +47,8 @@ public static ConfigDef getConfig() {
                 .define(KUSTO_AUTH_AUTHORITY, Type.STRING, null, Importance.HIGH, "Kusto auth using appid,appkey combo: authority")
                 .define(KUSTO_SINK_TEMPDIR, Type.STRING, System.getProperty("java.io.tempdir"), Importance.LOW, "Temp dir that will be used by kusto sink to buffer records. defaults to system temp dir")
                 .define(KUSTO_SINK_FLUSH_SIZE, Type.LONG, FileUtils.ONE_MB, Importance.HIGH, "Kusto sink max buffer size (per topic+partition combo)")
-                .define(KUSTO_SINK_FLUSH_INTERVAL_MS, Type.LONG, TimeUnit.MINUTES.toMillis(5), Importance.HIGH, "Kusto sink max staleness in milliseconds (per topic+partition combo)");
+                .define(KUSTO_SINK_FLUSH_INTERVAL_MS, Type.LONG, TimeUnit.MINUTES.toMillis(5), Importance.HIGH, "Kusto sink max staleness in milliseconds (per topic+partition combo)")
+                .define(KUSTO_AUTO_TABLE_CREATE, Type.BOOLEAN, false, Importance.LOW,"Mark true to allow connector to create the table if it not exists.");
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkConfig.java:51]: Changed username, password, appid, appkey, authority to ```Type.Password```

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 6]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
         return topicsToIngestionProps.get(topic);
     }
 
+    /**
+     *  This function validates whether the user has the read and write access to the intended table
+     *  before starting to sink records into ADX
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:161]: done

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 7]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+     * @param config config class as defined by the user.
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:162]: done

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 8]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+        ConnectionStringBuilder engineCsb =
+                ConnectionStringBuilder.createWithAadApplicationCredentials(
+                        config.getKustoUrl(),
+                        config.getKustoAuthAppid(),
+                        config.getKustoAuthAppkey(),
+                        config.getKustoAuthAuthority()
+                );
+        try {
+            Client engineClient = ClientFactory.createClient(engineCsb);
+            //To check whether the table and database exist
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:177]: done

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 9]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+            Results rs = engineClient.execute(database, table);
+            List<String> randomValueList = new ArrayList<>();
+            String ingestQuery = ".ingest inline into table %s <| %s";
+            //Creating random value liSt to be inserted into the test record
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:181]: done

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 10]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+            for (Entry entry : rs.getColumnNameToType().entrySet()) {
+                switch (entry.getValue().toString().toLowerCase()) {
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:183]: Will need to further investigate the cast type of each data type supported in Kusto. Will do

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 11]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+                    case "int32":
+                        randomValueList.add("1");
+                        break;
+                    default:
+                        break;
+                }
+            }
+            String randomValues = String.join(",", randomValueList);
+            rs = engineClient.execute(database, String.format(ingestQuery, table, randomValues));
+            // Extracting ExtentId to delete the temporary record.
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:197]: Will be using space,

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 12]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+            String extentId = rs.getValues().get(0).get(0);
+            //Deleting the temporary record.
+            engineClient.execute(database, String.format(".drop extent %s", extentId));
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:200]: Need to discuss on this.

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 13]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+            log.info("User has read and Write access to the table {}", table);
+        } catch (URISyntaxException e) {
+            throw new ConnectException("Unable to connect to Server", e);
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:203]: done

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java (hunk 1 above)
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:211]: Need to discuss

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 14]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+        } catch (DataClientException e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:205]: done

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 15]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+                        break;
+                }
+            }
+            String randomValues = String.join(",", randomValueList);
+            rs = engineClient.execute(database, String.format(ingestQuery, table, randomValues));
+            // Extracting ExtentId to delete the temporary record.
+            String extentId = rs.getValues().get(0).get(0);
+            //Deleting the temporary record.
+            engineClient.execute(database, String.format(".drop extent %s", extentId));
+            log.info("User has read and Write access to the table {}", table);
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:201]: done

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 16]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+
+        ConnectionStringBuilder engineCsb =
+                ConnectionStringBuilder.createWithAadApplicationCredentials(
+                        config.getKustoUrl(),
+                        config.getKustoAuthAppid(),
+                        config.getKustoAuthAppkey(),
+                        config.getKustoAuthAuthority()
+                );
+        try {
+            Client engineClient = ClientFactory.createClient(engineCsb);
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:176]: It doesn't have a close() method

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 17]
```diff
@@ -89,6 +102,8 @@ public static IngestClient createKustoIngestClient(KustoSinkConfig config) throw
                         String db = mapping.getString("db");
                         String table = mapping.getString("table");
 
+                        validateTableAccess(config, db, table);
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:105]: Moved the ```validateTableAccess()``` into a new function ```validateTableMapping()``` which validates all the table present in the *kusto.tables.topics_mapping* config. The ```validateTableMapping``` has been placed in the ```start``` function

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 18]
```diff
@@ -141,6 +156,68 @@ public TopicIngestionProperties getIngestionProps(String topic) {
...
+            engineClient.execute(database, String.format(".drop extent %s", extentId));
+            log.info("User has read and Write access to the table {}", table);
+        } catch (URISyntaxException e) {
+            throw new ConnectException("Unable to connect to Server", e);
+        } catch (DataClientException e) {
+            e.printStackTrace();
+        } catch (DataServiceException e) {
+            if (e.getCause().getMessage().contains("Database")) {
+                throw new ConfigException(
+                        String.format("Invalid or Denied Access to Database: %s", database), e);
```
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:226]: done

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (SanchayGupta1197) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:224]: done

**Pull request merged**